### PR TITLE
Result all the things

### PR DESCRIPTION
**Pull request by mltbnz**
This PR will introduce Swifts Result type to the network layer

### Before merging the PR

- [x] If applicable, does the PR contain enough unit / UI tests?
- [ ] If applicable, did you add a before / after screenshot of the feature?
- [ ] Did you update the [CHANGELOG.md](https://github.com/criticalmaps/criticalmaps-ios/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: CriticalMass/PostLocationRequest.swift [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+//  PostLocationRequest.swift
+//  CriticalMaps
+//
+//  Created by Malte Bünz on 16.08.19.
+//  Copyright © 2019 Pokus Labs. All rights reserved.
+//
+
+import Foundation
+
+struct PostLocationRequest: APIRequestDefining {
```
Comment (lennet) [inline CriticalMass/PostLocationRequest.swift]: similar to the comment about GetLocationsAndChatMessagesRequest

File: CriticalMass/ErrorHandler.swift [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+//  Created by Malte Bünz on 16.08.19.
+//  Copyright © 2019 Pokus Labs. All rights reserved.
+//
+
+import Foundation
+
+struct ErrorHandler {
+    static let `default` = ErrorHandler()
+
+    let genericErrorMessage = "Sorry, something went wrong"
```
Comment (lennet) [inline CriticalMass/ErrorHandler.swift]: this need to be localised

File: CriticalMass/ErrorHandler.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    func handleError(_ error: Error?) {
+        if let errorDescription = error?.localizedDescription {
+            presentToUser(message: errorDescription)
+        } else {
+            presentToUser(message: genericErrorMessage)
+        }
+    }
+
+    private func presentToUser(message: String) {
+        print(message)
```
Comment (lennet) [inline CriticalMass/ErrorHandler.swift]: Do we want to add alerts in this case? Although we shouldn't present alerts if the app is running in the background

File: CriticalMass/NetworkOperator.swift [hunk 4]
```diff
@@ -18,36 +18,55 @@ struct NetworkOperator: NetworkLayer {
...
-        var request = URLRequest(url: url)
-        request.httpMethod = "POST"
-        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
-        request.httpBody = bodyData
-        dataTask(with: request, decodable: decodable, customDateFormatter: nil, completion: completion)
+    func post<T: APIRequestDefining>(request: T, bodyData: Data, completion: @escaping ResultCallback<T.ResponseDataType>) {
+        var urlRequest = request.makeRequest()
+        urlRequest.httpBody = bodyData
+        dataTask(with: urlRequest) { result in
+            switch result {
```
Comment (lennet) [inline CriticalMass/NetworkOperator.swift]: minor: there is some repetition in the result handling between post and get. I'm wondering if we could simplify that

File: CriticalMaps.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -1028,7 +1092,7 @@
 				CLANG_ENABLE_MODULES = YES;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
-				DEVELOPMENT_TEAM = 5YLLXUZBZ2;
+				DEVELOPMENT_TEAM = BP8836JVQR;
```
[the remark this answers is not in the document]
Comment (mltbnz) [reply CriticalMaps.xcodeproj/project.pbxproj]: Right. I missed that

File: CriticalMass/GetLocationsAndChatMessagesRequest.swift [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+//  GetLocationsAndChatMessagesRequest.swift
+//  CriticalMaps
+//
+//  Created by Malte Bünz on 11.08.19.
+//  Copyright © 2019 Pokus Labs. All rights reserved.
+//
+
+import Foundation
+
+struct GetLocationsAndChatMessagesRequest: APIRequestDefining {
```
[the remark this answers is not in the document]
Comment (mltbnz) [reply CriticalMass/GetLocationsAndChatMessagesRequest.swift]: I did not find a solution with the protocol to add a default here. Do you know a way?

File: CriticalMass/NetworkOperator.swift (hunk 4 above)
Comment (mltbnz) [reply CriticalMass/NetworkOperator.swift]: Good point. I was thinking of something like passing in the body as an optional and then only adding it if it has a value

File: CriticalMass/GetLocationsAndChatMessagesRequest.swift [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    typealias ResponseDataType = ApiResponse
+    var baseUrl: URL
+    var paths: [String]
+    var headers: HTTPHeaders?
```
Comment (lennet) [inline CriticalMass/GetLocationsAndChatMessagesRequest.swift]:
```suggestion
    var headers: HTTPHeaders? = nil
```

File: CriticalMass/GetLocationsAndChatMessagesRequest.swift [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+//  Created by Malte Bünz on 11.08.19.
+//  Copyright © 2019 Pokus Labs. All rights reserved.
+//
+
+import Foundation
+
+struct GetLocationsAndChatMessagesRequest: APIRequestDefining {
+    typealias ResponseDataType = ApiResponse
+    var baseUrl: URL
+    var paths: [String]
```
Comment (lennet) [inline CriticalMass/GetLocationsAndChatMessagesRequest.swift]:
```suggestion
    var paths: [String] = []
```

File: CriticalMass/GetLocationsAndChatMessagesRequest.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (lennet) [reply CriticalMass/GetLocationsAndChatMessagesRequest.swift]: added two example suggestions

File: CriticalMass/ErrorHandler.swift (hunk 3 above)
Comment (mltbnz) [reply CriticalMass/ErrorHandler.swift]: I want to add something visual after this PR is merged. For now I am not sure how it looks like. Maybe also @zutrinken has an idea of error states

File: CriticalMass/NetworkOperator.swift (hunk 4 above)
Comment (mltbnz) [reply CriticalMass/NetworkOperator.swift]: Oh. I confused your reply. I did not find a good (quick) solution. I want to find something tho and will park the ticket some time until I can work on a solution

File: CriticalMass/ErrorHandler.swift (hunk 2 above)
Comment (mltbnz) [reply CriticalMass/ErrorHandler.swift]: I want to add something visual after this PR is merged. For now I am not sure how it looks like. Maybe also @zutrinken has an idea of error states

File: CriticalMass/GetLocationsAndChatMessagesRequest.swift (hunk 8 above)
Comment (mltbnz) [reply CriticalMass/GetLocationsAndChatMessagesRequest.swift]: I added a struct `Endpoint` that solves the issue I think. I also updated the creation logic of the request

**Pull request merged**